docs(tree): general schema evolution docs (#25218)
microsoft/FluidFramework: docs/docs/data-structures/tree/schema-evolution/index.mdx

Reviewer: Why mdx instead of md?

Author: Almost all the pages in the folder are mdx so I decided to just copy that. It's probably not necessary and maybe everyone was thinking that and that's how we got to this point but I figured it doesn't hurt. 

@Josmithr should we be keeping to using md files on the site?

Reviewer 2: `mdx` is probably preferred when possible. The only remaining reason to use plain `md` at this point is if you are using `markdown-magic` to embed contents from another file in the repo in your document - that doesn't support mdx yet.

---
title: Schema Evolution
sidebar_position: 8
---

# Schema Evolution

As your application grows and evolves, you may need to modify the structure of your [SharedTree](../index.mdx) data.

> SharedTree provides safe, backward-compatible ways to update your [schemas](../schema-definition.mdx) without breaking existing data or older application versions (see [understanding compatibility](./index.mdx#understanding-compatibility)).

Reviewer 3: ```suggestion
SharedTree makes it possible to update your [schemas](../schema-definition.mdx) in ways that can be leveraged to ensure backward compatibility and forward compatibility (see [understanding compatibility](./index.mdx#understanding-compatibility)).
```
"without breaking existing data" sounds like a document corruption risk, which is not really what we're concerned with here. I'm guessing meant breaking new application's ability to work on existing data, which is redundant with the "backward-compatible" earlier in the sentence. 

(Note: reading further down I'm noticing that you and I are not on the same page w.r.t. to the meaning of backward compatibility and forward compatibility. I'm guessing this is how you ended up with something that sounds redundant to me but probably not to you).

My suggestion also shifts the language from "SharedTree provides safe [...] ways" to "ways that can be leveraged to ensure...". This is important because we can't make the claim that using SharedTree APIs ensures compatibility. It's how they're used that determines whether compatibility is ensured.

---
title: Schema Evolution
sidebar_position: 8
---

# Schema Evolution

As your application grows and evolves, you may need to modify the structure of your [SharedTree](../index.mdx) data.
SharedTree makes it possible to update your [schemas](../schema-definition.mdx) in ways that allow an app to load or upgrade documents created or edited by earlier versions of the app (see [understanding compatibility](./index.mdx#understanding-compatibility)).

> > **Note:** For detailed information about which types of changes are safe versus breaking, see [types of schema changes](./types-of-changes).

Reviewer 2: This seems like a good use case for [admonition](https://docusaurus.io/docs/markdown-features/admonitions) syntax :)

Reviewer 2: Same for the new note in `schema-definition`.

SharedTree makes it possible to update your [schemas](../schema-definition.mdx) in ways that allow an app to load or upgrade documents created or edited by earlier versions of the app (see [understanding compatibility](./index.mdx#understanding-compatibility)).

> **Note:** For detailed information about which types of changes are safe versus breaking, see [types of schema changes](./types-of-changes).

## Stored vs View Schema

When updating your schema, it's important to understand the difference between the two types of schemas:

- **View Schema**: The schema defined in your application code using [`SchemaFactory`](../../../api/fluid-framework/schemafactory-class)
- **Stored Schema**: The schema persisted with each document. It describes what type of tree can be stored in the document.

When a document is first created, its stored schema is set using the given view schema.
Afterwards, it is immutable until explicitly [upgraded](./index.mdx#schema-upgrade-process).

### Stored Schema Invariants

The stored schema follows additional constraints that don't apply to view schemas:

**1. Complete Description**

> The stored schema completely describes all data in the tree.

Reviewer: ```suggestion
The stored schema completely describes all possible data in the tree.
```

Every node and field in the document is known to the stored schema.

**2. Additive Only**
The stored schema can grow over time but cannot shrink:
- New allowed types can be added to fields
- Existing types cannot be removed from fields

> - Fields cannot be removed from node types

Reviewer: ```suggestion
- Fields cannot be removed from object nodes
```

> - Node types cannot be completely removed

Reviewer: This is potentially useful information for someone trying to deeply understand how the system works, but is probably unhelpful and confusing to someone trying to learn how to do a schema migration for the first time. Node types cannot be _completely_ removed? What does that mean - can they remove nodes? Or not? Where? When?

Maybe instead, say something like "All nodes types present in the stored schema will exist forever, even if they stop being used by the view schema"

**3. Field Key Stability**

> Field keys are permanent once introduced.

Reviewer: ```suggestion
Field keys on object nodes are permanent once introduced.
```

> If a field is removed and later reintroduced with a different type, SharedTree cannot distinguish between old and new data without inspecting values.

Reviewer: ```suggestion
Therefore, a field key on an object node in the stored schema can never be changed/reused/repurposed. 
```

### Why These Constraints?

It's assumed that a Fluid application developer does not have absolute control over all user documents.
Documents can persist for an unbounded length of time before they are re-opened by an application.
Therefore, a Fluid application should be prepared to encounter any version of a document, even a very old one, and handle it appropriately.

## Schema Compatibility

When you open a document, SharedTree compares your application's view schema with the document's stored schema to determine compatibility:

- **Compatible**: Document can be opened and used normally
- **Upgradeable**: Document can be upgraded to match your view schema using [`view.upgradeSchema()`](../../../api/fluid-framework/treeview-interface#upgradeschema-method)

> - **Incompatible**: Document cannot be opened with your current view schema

Reviewer: ```suggestion
- **Incompatible**: Document cannot be opened with or upgraded to match your current view schema
```

They must be compatible or upgradeable in order for the application to work - otherwise, it cannot read or write data from/to the document.
SharedTree will throw an assert if the application attempts to access the document's data with an incompatible view schema.

> Initially, a document is always compatible with the application that created it (i.e. the view schema is compatible with the stored schema).

Reviewer: ```suggestion
Initially, a document's stored schema is created to match the view schema at the time of creation, and therefore the view schema is compatible with the stored schema.
```

#### 😎 Chill
| Backwards Compatible | Forwards Compatible | Rollout Process |
|-----------------------|----------------------|-----------------|
| ✅                    | ✅                   | Normal          |

**Examples:**
- Adding an optional field when compatibility flag is enabled

#### 🌶️ Spicy
| Backwards Compatible | Forwards Compatible | Rollout Process |
|-----------------------|----------------------|-----------------|
| ✅                    | ❌                   | ⚠️ Staged       |

**Examples:**
- Adding a new allowed type (TODO link to how to doc)

#### 👮 Go to Jail
| Backwards Compatible | Forwards Compatible | Rollout Process |
|-----------------------|----------------------|-----------------|
| ❌                    | ❌                   | 🚫 Forbidden    |

**Examples:**
- Adding/removing a required field
- Removing an allowed type from a field
- Renaming a node identifier

### Checking Compatibility

Use [`TreeView.compatibility`](../../../api/fluid-framework/treeview-interface#compatibility-property) to check if your schema changes are compatible with existing documents:

```typescript
const view = tree.viewWith(configuration);
const { compatibility } = view;

> if (!compatibility.isEquivalent) {

Reviewer: I think we should just stop advertising `isEquivalent`. It doesn't really do anything sensible. It's safe to call `upgradeSchema()` even if nothing has changed - we will no-op and no harm done.

    if (compatibility.canUpgrade) {
        view.upgradeSchema(); // Safe to upgrade
    } else {
        // Handle incompatible schema
        throw new Error("Cannot open document with current schema");
    }
}
```

### Monitoring Remote Schema Changes

In collaborative applications, other clients may upgrade the document's schema while you're working.
Listen for the `schemaChanged` event to handle these remote upgrades:

```typescript
// Monitor for remote schema changes
view.events.on("schemaChanged", () => {

>     if (!view.compatibility.canView) {

Reviewer: What about if the schema changes and `canView` is falso but `canUpgrade` is true? Is the comment below implying that if the user reloads, then the upgrade will be done upon reloading?

        // Document was upgraded to an incompatible schema
        // Handle gracefully - perhaps reload or notify user
        throw new Error("Document was upgraded and is no longer compatible");
    }

    // Schema changed but is still compatible
    // You may need to refresh UI or handle new data structures
    console.log("Document schema was updated by another client");
});
```

## Schema Upgrade Process

> When you make [backwards compatible](./index.mdx#understanding-compatibility) changes to your schema, you can upgrade existing documents to use the new schema using [`view.upgradeSchema()`](../../../api/fluid-framework/treeview-interface#upgradeschema-method):

Reviewer: This makes it sound like we're introducing `upgradeSchema()` for the first time, but we already showed it in the example above. That's probably okay if we don't expect users to read this doc from top to bottom. But I do think that they need to have read the example above before they read this one - understanding the compatibility API is a prerequisite to understanding the upgrade API.

Reviewer: That's probably what I would have done too 👍

const factory = new SchemaFactory("WhiteboardApp");

// Updated schema with new optional field
class Note extends factory.object("Note", {
    id: factory.string,
    x: factory.number,
    y: factory.number,
    color: factory.optional(factory.string), // New optional field
}) {}

// Create tree view with updated schema
const config = new TreeViewConfiguration({ schema: Note });
const view = tree.viewWith(config);

// Check compatibility and upgrade if needed
if (!view.compatibility.isEquivalent) {
    if (view.compatibility.canUpgrade) {
        view.upgradeSchema(); // Upgrade document to new schema
        console.log("Document upgraded successfully");
    } else {
        throw new Error("Schema is incompatible and cannot be upgraded");
    }
}

// Now you can use the upgraded document
view.root.color = "#FF0000";
```

For detailed information about which types of changes are safe versus breaking, see [Types of Schema Changes](./types-of-changes).

### Staged Rollouts

When deploying schema changes in production, use staged rollouts to ensure compatibility between clients running different application versions.

**Why Staged Rollouts Are Necessary**

> 

Reviewer: Let's emphasize here that staged rollouts allow you to handle changes that are backwards compatible but not forwards compatible, to connect it with the sections above.

In collaborative applications, multiple users with different client versions may work on the same document simultaneously.
If a client upgrades a document to a new schema that older clients cannot read, those users will be locked out until they update their application.

**Two-Phase Deployment Strategy**

**Phase 1: Deploy Schema Reading Support**
- Deploy application versions that can read the new schema without upgrading documents
- Wait until all (or "tolerably most") users have updated to the new application version. This is called **client saturation**.
- Monitor deployment metrics to ensure coverage

**Phase 2: Enable Document Upgrades**
- Deploy code that automatically upgrades documents to the new schema
- All clients can now collaborate since they support reading the upgraded schema
- Monitor for any compatibility issues

> Theoretically, some changes don't require a staged rollout and are therefore capable of cross-version collaboration.

Reviewer: Specifically, changes that are both backwards _and_ forwards compatible.

In that case, new clients can read and write to the same document as old clients without either one being fundamentally incompatible.

> However, staged rollouts are not easily avoidable. Even seemingly simple schema changes, like adding a single required field to a node, cannot support cross-version collaboration and require a staged rollout.

Reviewer: ```suggestion
However, staged rollouts are not easily avoidable. Even seemingly simple schema changes (for example, adding a new type of node to a field) may not be forwards-compatible, cannot support cross-version collaboration and therefore require a staged rollout.
```

Adding required fields is a confusing example because it's not possible except it is possible in a roundabout way. Probably a detail to be saved for the type-of-changes document.
